### PR TITLE
ioutils: add MultiFileReader to ease concatenation of multiple readers

### PR DESCRIPTION
**Pull request by durin42**
I ended up needing something like this for Mercurial, and mhashemirc
suggested that it would make sense in boltons.

### CODE REVIEW

**Comment by mahmoud** [comment]
Looks great, @durin42, thanks! I'll handle the 2.6 compat, it's nbd. I'll also take a swing at getting docs in there, too. I'll @ you if there's questions.

**Comment by mahmoud** [comment]
So, @durin42, while writing some docs on this, it occurred to me that we only take `BytesIO` and `TextIOBase`, and for Python 2 we should probably take `file` objects (since that's the class name and `__init__` arg name, after all). Does it sound ok to wrap up passed-in `file` objects with a `BytesIO`?

**Comment by induane** [comment]
@durin42 this looks pretty awesome btw!

@mahmoud  It might make sense to accept SpookedBytesIO and SpooledStringIO instances in the multifile reader as well. It might be confusing to disallow them given that they're file-like-objects in the same module.

**Comment by durin42** [comment]
@mahmoud Hmm, file objects can be open in either bytes or unicode mode in Python 3. Sniff the mode attribute and look for a 'b' to verify it's bytes mode, otherwise it's reading unicodes.

It should be fine to not have to wrap the files at all - just verify they'll return the right kind of data and then leave them in the right place in the sequence (assuming they support seek).

Also agree that handling the Spooled{Bytes,String}IO instances seems reasonable. I didn't think of that because I didn't have such a requirement in hg. :)

File: tests/test_urlutils.py [hunk 1]
```diff
@@ -313,6 +313,13 @@ def test_navigate():
     assert navd.to_text() == _dest_text
 
 
+# TODO: RFC3986 6.2.3 (not just for query add, either)
```
Comment (durin42) [inline tests/test_urlutils.py]: Wrong branch?

Comment (mahmoud) [inline tests/test_urlutils.py]: heh, galdangit.

**Comment by mahmoud** [comment]
@durin42 On Python3, `open` returns `io` objects:

```
>>> isinstance(open('test.md'), io.TextIOBase)
True
```

Interestingly, even with 'rb', we get a `TextIOBase`:

```
>>> isinstance(open('test.md', 'rb'), io.TextIOBase)
True
>>> isinstance(open('test.md', 'rb'), io.BytesIO)
False
```

So, depending on your definition, it should "just work" for Python 3. It's mostly handling `file` on py2, which always returns bytes.

**Comment by durin42** [comment]
Ugh, feels like a Python3 bug that a binary-mode file isinstance TextIOBase.

On Python 2, yeah, this is already solved. On Python 3, I guess we have to look for a "mode" attr and then sniff it to figure out if it's a bytes-like or text-like file.

:(

**Comment by mahmoud** [comment]
@durin42 gotcha, all fixed. i think this is basically ready to merge. Added some docs, too. A quick glance at those would be appreciated, too.

**Comment by mahmoud** [comment]
@induane the current version also supports `Spooled*IO`, but I should note that not taking an `initial_value` in the constructor, and having the string first argument silently becoming `max_size` threw me for a bit of a loop. It'd be nice to get a PR to address that, but we can discuss that in a separate issue! :)

**Comment by durin42** [comment]
Docs look good to me. Thanks for finishing off my prototype!

**Pull request merged**